File: aribook/_template_aribook237DP分割数.cpp
// aribook237 DP 分割数
#include<iostream>
#include<algorithm>
//#include <numeric>
//#include<string>
//#include<vector>
//#include<map>
//#include<tuple>
//#include<queue>
//#include<regex>

// #include <bits/stdc++.h>
#define rep(i,n) for (int i = 0; i < (n); ++i)
using namespace std;
typedef long long ll;

const int MAX_N = 1000;
const int MAX_M = 1000;

// 入力
int n, m;
int MOD = 10000;
int dp[MAX_M + 1][MAX_N + 1];	// DPテーブル

int main() {
	cin >> n >> m;

	// 動的計画法(DP)
	dp[0][0] = 1;
	for (int i = 1; i < m + 1; i++) {
		rep (j, n + 1) {
			if (j - i >= 0) {
				dp[i][j] = (dp[i - 1][j] + dp[i][j - i]) % MOD;
			}
			else {
				dp[i][j] = dp[i - 1][j];
			}
		}
	}

	cout << dp[m][n] << endl;

	return 0;
}

File: aribook/_template_aribook237DP分割数.py
# aribook237 DP 分割数

import numpy as np

MAX_N = 1000
MAX_M = 1000
MOD = 10000

N, M = map(int, input().split())

dp = np.zeros((M+1, N+1), np.int)   # DPテーブル

dp[0, 0] = 1
for i in range(1, M+1):
    for j in range(N+1):
        if j -i  >= 0:
            dp[i, j] = (dp[i-1, j] + dp[i, j-i]) % MOD
        else:
            dp[i, j] = dp[i-1, j]

print(dp[M, N])

"""
input
N M

testdata
4 3

output
4
"""
